File: final/src/grammar/Parser.java
package grammar;

import java.io.FileInputStream;
import java.io.IOException;
import java.util.LinkedHashSet;

import errors.SyntaxError;
import lexer.Lexer;
import lexer.Tag;
import lexer.Token;

public class Parser {
	private Lexer lexer;
	private Token token;
	private static LinkedHashSet<Integer> primaryExpressionFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> unaryExpressionFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> multiplicativeExpressionFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> additiveExpressionFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> relationalExpressionFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> equalityExpressionFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> conditionalTermFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> conditionalExpressionFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> expressionFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> drawingStatementFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> movementStatementFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> simpleStatementFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> structuredStatementFirst = new LinkedHashSet<Integer>();
	private static LinkedHashSet<Integer> statementFirst = new LinkedHashSet<Integer>();

	public Parser (FileInputStream input) {
		lexer = new Lexer(input);
		
		primaryExpressionFirst.add(Tag.ID); primaryExpressionFirst.add(Tag.NUMBER);
		primaryExpressionFirst.add(Tag.TRUE); primaryExpressionFirst.add(Tag.FALSE);
		primaryExpressionFirst.add((int) '('); 
		
		unaryExpressionFirst.add((int) '-'); unaryExpressionFirst.add(Tag.NOT);
		unaryExpressionFirst.addAll(primaryExpressionFirst);
		
		multiplicativeExpressionFirst.addAll(unaryExpressionFirst);
		
		additiveExpressionFirst.addAll(multiplicativeExpressionFirst);
		
		relationalExpressionFirst.addAll(additiveExpressionFirst);
		
		equalityExpressionFirst.addAll(relationalExpressionFirst);
		
		conditionalTermFirst.addAll(equalityExpressionFirst);
		
		conditionalExpressionFirst.addAll(conditionalTermFirst);
		
		expressionFirst.addAll(conditionalExpressionFirst);
		
		drawingStatementFirst.add(Tag.CLEAR); drawingStatementFirst.add(Tag.CIRCLE);
		drawingStatementFirst.add(Tag.ARC); drawingStatementFirst.add(Tag.PENUP);
		drawingStatementFirst.add(Tag.PENDOWN); drawingStatementFirst.add(Tag.COLOR);
		drawingStatementFirst.add(Tag.PENWIDTH);
		
		movementStatementFirst.add(Tag.FORWARD); movementStatementFirst.add(Tag.BACKWARD);
		movementStatementFirst.add(Tag.LEFT); movementStatementFirst.add(Tag.RIGHT);
		movementStatementFirst.add(Tag.SETX); movementStatementFirst.add(Tag.SETY);
		movementStatementFirst.add(Tag.SETXY); movementStatementFirst.add(Tag.HOME);
		
		simpleStatementFirst.add(Tag.MAKE); simpleStatementFirst.addAll(movementStatementFirst);
		simpleStatementFirst.addAll(drawingStatementFirst); simpleStatementFirst.add(Tag.PRINT);
		
		structuredStatementFirst.add(Tag.IF); structuredStatementFirst.add(Tag.IFELSE);
		structuredStatementFirst.add(Tag.REPEAT);
		
		statementFirst.addAll(simpleStatementFirst); statementFirst.addAll(structuredStatementFirst);
	}

	private void check(int tag) throws SyntaxError, IOException {
		if (token.getTag() == tag) {
			token = lexer.scan();
		} else {
			throw new SyntaxError();
		}
	}

	public void analyze() throws SyntaxError, IOException {
		token = lexer.scan();
		program();
		if (token.getTag() == Tag.EOF) {
			System.out.println("ACCEPTED");
		} else {
			throw new SyntaxError();
		}
	}
	
	private void primaryExpression() throws SyntaxError, IOException {
		if (token.getTag() == Tag.ID) {
			check(Tag.ID);
		} else if (token.getTag() == Tag.NUMBER) {
			check(Tag.NUMBER);
		} else if (token.getTag() == Tag.TRUE) {
			check(Tag.TRUE);
		} else if (token.getTag() == Tag.FALSE) {
			check(Tag.FALSE);
		} else if (token.getTag() == ((int) '(')) {
			check((int) '(');
			expression();
			check((int) ')');
		} else {
			throw new SyntaxError();
		}
	}
	
	private void unaryExpression() throws SyntaxError, IOException {
		if (token.getTag() == ((int) '-')) {
			check((int) '-');
			unaryExpression();
		} else if (token.getTag() == Tag.NOT) {
			check(Tag.NOT);
			unaryExpression();
		} else if (primaryExpressionFirst.contains(token.getTag())) {
			primaryExpression();
		} else {
			throw new SyntaxError();
		}
	}
	
	private void extendedMultiplicativeExpression() throws SyntaxError, IOException {
		if (token.getTag() == ((int) '*')) {
			check((int) '*');
			unaryExpression();
			extendedMultiplicativeExpression();
		} else if (token.getTag() == ((int) '/')) {
			check((int) '/');
			unaryExpression();
			extendedMultiplicativeExpression();
		} else if (token.getTag() == Tag.MOD) {
			check(Tag.MOD);
			unaryExpression();
			extendedMultiplicativeExpression();
		} else {
			// do nothing
		}
	}
	
	private void multiplicativeExpression() throws SyntaxError, IOException {
		if (unaryExpressionFirst.contains(token.getTag())) {
			unaryExpression();
			extendedMultiplicativeExpression();
		} else {
			throw new SyntaxError();
		}
	}
	
	private void extendedAdditiveExpression() throws SyntaxError, IOException {
		if (token.getTag() == ((int) '+')) {
			check((int) '+');
			multiplicativeExpression();
			extendedAdditiveExpression();
		} else if (token.getTag() == ((int) '-')) {
			check((int) '-');
			multiplicativeExpression();
			extendedAdditiveExpression();
		} else {
			// do nothing
		}
	}
	
	private void additiveExpression() throws SyntaxError, IOException {
		if (multiplicativeExpressionFirst.contains(token.getTag())) {
			multiplicativeExpression();
			extendedAdditiveExpression();
		} else {
			throw new SyntaxError();
		}
	}
	
	private void extendedRelationalExpression() throws SyntaxError, IOException {
		if (token.getTag() == ((int) '<')) {
			check((int) '<');
			additiveExpression();
			extendedRelationalExpression();
		} else if (token.getTag() == Tag.LEQ) {
			check(Tag.LEQ);
			additiveExpression();
			extendedRelationalExpression();
		} else if (token.getTag() == ((int) '>')) {
			check((int) '>');
			additiveExpression();
			extendedRelationalExpression();
		} else if (token.getTag() == Tag.GEQ) {
			check(Tag.GEQ);
			additiveExpression();
			extendedRelationalExpression();
		} else {
			// do nothing
		}
	}
	
	private void relationalExpression() throws SyntaxError, IOException {
		if (additiveExpressionFirst.contains(token.getTag())) {
			additiveExpression();
			extendedRelationalExpression();
		} else {
			throw new SyntaxError();
		}
	}
	
	private void extendedEqualityExpression() throws SyntaxError, IOException {
		if (token.getTag() == ((int) '=')) {
			check((int) '=');
			relationalExpression();
			extendedEqualityExpression();
		} else if (token.getTag() == Tag.NEQ) {
			check(Tag.NEQ);
			relationalExpression();
			extendedEqualityExpression();
		} else {
			// do nothing
		}
	}
	
	private void equalityExpression() throws SyntaxError, IOException {
		if (relationalExpressionFirst.contains(token.getTag())) {
			relationalExpression();
			extendedEqualityExpression();
		} else {
			throw new SyntaxError();
		}
	}
	
	private void extendedConditionalTerm() throws SyntaxError, IOException {
		if (token.getTag() == Tag.AND) {
			check(Tag.AND);
			equalityExpression();
			extendedConditionalTerm();
		} else {
			// do nothing
		}
	}
	
	private void conditionalTerm() throws SyntaxError, IOException {
		if (equalityExpressionFirst.contains(token.getTag())) {
			equalityExpression();
			extendedConditionalTerm();
		} else {
			throw new SyntaxError();
		}
	}
	
	private void extendedConditionalExpression() throws SyntaxError, IOException {
		if (token.getTag() == Tag.OR) {
			check(Tag.OR);
			conditionalTerm();
			extendedConditionalExpression();
		} else {
			// do nothing
		}
	}
	
	private void conditionalExpression() throws SyntaxError, IOException {
		if (conditionalTermFirst.contains(token.getTag())) {
			conditionalTerm();
			extendedConditionalExpression();
		} else {
			throw new SyntaxError();
		}
	}
	
	private void expression() throws SyntaxError, IOException {
		if (conditionalExpressionFirst.contains(token.getTag())) {
			conditionalExpression(); 
		} else {
			throw new SyntaxError();
		}
	}
	
	private void ifElseStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.IFELSE) {
			check(Tag.IFELSE);
			expression();
			check((int) '[');
			statementSequence();
			check((int) ']');
			check((int) '[');
			statementSequence();
			check((int) ']');
		} else {
			throw new SyntaxError();
		}
	}
	
	private void ifStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.IF) {
			check(Tag.IF);
			expression();
			check((int) '[');
			statementSequence();
			check((int) ']');
		} else {
			throw new SyntaxError();
		}
	}
	
	private void conditionalStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.IF) {
			ifStatement();
		} else if (token.getTag() == Tag.IFELSE) {
			ifElseStatement();
		} else {
			throw new SyntaxError();
		}
	}
	
	private void repeatStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.REPEAT) {
			check(Tag.REPEAT);
			expression();
			check((int) '[');
			statementSequence();
			check((int) ']');
		} else {
			throw new SyntaxError();
		}
	}
	
	private void structuredStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.IF || token.getTag() == Tag.IFELSE) {
			conditionalStatement();
		} else if (token.getTag() == Tag.REPEAT) {
			repeatStatement();
		} else {
			throw new SyntaxError();
		}
	}
	
	
	private void moreElements() throws SyntaxError, IOException {
		if (token.getTag() == ((int) ',')) {
			check((int) ',');
			element();
			moreElements();
		} else {
			//do nothing
		}
	}
	
	private void element() throws SyntaxError, IOException {
		if (token.getTag() == Tag.STRING) {
			check(Tag.STRING);
		} else if (expressionFirst.contains(token.getTag())) {
			expression();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void textStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.PRINT) {
			check(Tag.PRINT);
			check((int) '[');
			element();
			moreElements();
			check((int) ']');
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void penWidthStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.PENWIDTH) {
			check(Tag.PENWIDTH);
			expression();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void colorStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.COLOR) {
			check(Tag.COLOR);
			expression();
			expression();
			expression();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void penDownStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.PENDOWN) {
			check(Tag.PENDOWN);
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void penUpStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.PENUP) {
			check(Tag.PENUP);
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void arcStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.ARC) {
			check(Tag.ARC);
			expression();
			expression();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void circleStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.CIRCLE) {
			check(Tag.CIRCLE);
			expression();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void clearStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.CLEAR) {
			check(Tag.CLEAR);
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void drawingStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.CLEAR) {
			clearStatement();
		} else if (token.getTag() == Tag.CIRCLE) {
			circleStatement();
		} else if (token.getTag() == Tag.ARC) {
			arcStatement();
		} else if (token.getTag() == Tag.PENUP) {
			penUpStatement();
		} else if (token.getTag() == Tag.PENDOWN) {
			penDownStatement();
		} else if (token.getTag() == Tag.COLOR) {
			colorStatement();
		} else if (token.getTag() == Tag.PENWIDTH) {
			penWidthStatement();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void homeStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.HOME) {
			check(Tag.HOME);
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void setXYStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.SETXY) {
			check(Tag.SETXY);
			expression();
			expression();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void setYStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.SETY) {
			check(Tag.SETY);
			expression();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void setXStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.SETX) {
			check(Tag.SETX);
			expression();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void rightStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.RIGHT) {
			check(Tag.RIGHT);
			expression();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void leftStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.LEFT) {
			check(Tag.LEFT);
			expression();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void backwardStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.BACKWARD) {
			check(Tag.BACKWARD);
			expression();
		} else {
			throw new SyntaxError();
		} 
	}
	
	private void forwardStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.FORWARD) {
			check(Tag.FORWARD);
			expression();
		} else {
			throw new SyntaxError();
		}		
	}
	
	private void movementStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.FORWARD) {
			forwardStatement();
		} else if (token.getTag() == Tag.BACKWARD) {
			backwardStatement();
		} else if (token.getTag() == Tag.LEFT) {
			leftStatement();
		} else if (token.getTag() == Tag.RIGHT) {
			rightStatement();
		} else if (token.getTag() == Tag.SETX) {
			setXStatement();
		} else if (token.getTag() == Tag.SETY) {
			setYStatement();
		} else if (token.getTag() == Tag.SETXY) {
			setXYStatement();
		} else if (token.getTag() == Tag.HOME) {
			homeStatement();
		} else {
			throw new SyntaxError();
		}		
	}
	
	private void assigmentStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.MAKE) {
			check(Tag.MAKE);
			check(Tag.ID);
			expression();
		} else {
			throw new SyntaxError();
		}		
	}
	
	private void simpleStatement() throws SyntaxError, IOException {
		if (token.getTag() == Tag.MAKE) {
			assigmentStatement();
		} else if (movementStatementFirst.contains(token.getTag())) {
			movementStatement();
		} else if (drawingStatementFirst.contains(token.getTag())) {
			drawingStatement();
		} else if (token.getTag() == Tag.PRINT) {
			textStatement();
		} else {
			throw new SyntaxError();
		}
	}
	
	
	private void statement() throws SyntaxError, IOException {
		if (simpleStatementFirst.contains(token.getTag())) {
			simpleStatement();
		} else if (structuredStatementFirst.contains(token.getTag())) {
			structuredStatement();
		} else {
			throw new SyntaxError();
		}
	}
	
	private void statementSequence() throws SyntaxError, IOException {
		if (statementFirst.contains(token.getTag())) {
			statement();
			statementSequence();
		} else {
			// do nothing
		}
	}
	
	private void program() throws SyntaxError, IOException {
		if (statementFirst.contains(token.getTag())) {
			statementSequence();
		} else {
			throw new SyntaxError();
		}
	}
}

File: stage-3/Parser.py
from Lexer import *
from Environment import *

class Parser:
    __lex = None
    __token = None
    __env = None
    __output = ""

    def __init__(self, filepath):
        self.__lex = Lexer(filepath)
        self.__token = None
        self.__env = Environment()
        self.__output = ""

        self.__firstPrimaryExpression = set((Tag.ID, Tag.NUMBER, Tag.TRUE, Tag.FALSE, ord('(')))

        self.__firstUnaryExpression = self.__firstPrimaryExpression.union( set((ord('-'), ord('!'))) )

        self.__firstExtendedMultiplicativeExpression = set((ord('*'), ord('/'), Tag.MOD))

        self.__firstMultiplicativeExpression = self.__firstUnaryExpression

        self.__firstExtendedAdditiveExpression = set((ord('+'), ord('-')))

        self.__firstAdditiveExpression = self.__firstMultiplicativeExpression

        self.__firstExtendedRelationExpresion = set((ord('<'), Tag.LEQ, ord('>'), Tag.GEQ))

        self.__firstRelationalExpression = self.__firstAdditiveExpression

        self.__firstExtendedEqualityExpression = set((ord('='), Tag.NEQ))

        self.__firstEqualityExpression = self.__firstRelationalExpression

        self.__firstConditionalTerm = self.__firstEqualityExpression

        self.__firstConditionalExpression = self.__firstConditionalTerm

        self.__firstExpression = self.__firstConditionalExpression

        self.__firstConditionalStatement = set((Tag.IF, Tag.IFELSE))

        self.__firstStructuredStatement = self.__firstConditionalStatement.union({Tag.REPEAT})

        self.__firstElement = self.__firstExpression.union({Tag.STRING})

        self.__firstDrawingStatement = set((Tag.CLEAR, Tag.CIRCLE, Tag.ARC, Tag.PENUP, 
                                            Tag.PENDOWN, Tag.COLOR, Tag.PENWIDTH))

        self.__firstMovementStatement = set((Tag.FORWARD, Tag.BACKWARD, Tag.RIGHT, Tag.LEFT, 
                                             Tag.SETX, Tag.SETY, Tag.SETXY, Tag.HOME))
        
        self.__firstSimpleStatement = (self.__firstMovementStatement.union(self.__firstDrawingStatement)).union(
            set((Tag.VAR, Tag.ID, Tag.PRINT)))
        
        self.__firstStatement = self.__firstSimpleStatement.union(self.__firstStructuredStatement)

        self.__firstStatementSequence = self.__firstStatement

        self.__firstProgram = self.__firstStatementSequence

    def __check(self, tag, extra = None):
        if self.__token.getTag() == tag:
            self.__token = self.__lex.scan()
        else:
            text = 'Line ' + str(self.__lex.getLine())
            if extra == None:
                text = text + "."
            else:
                text = text + extra
            raise Exception(text)
    
    def analize(self):
        self.__token = self.__lex.scan()
        self.__program()

    def __primaryExpression(self):
        if self.__token.getTag() in self.__firstPrimaryExpression:
            if self.__token.getTag() == Tag.ID:
                self.__check(Tag.ID)
            elif self.__token.getTag() == Tag.NUMBER:
                self.__check(Tag.NUMBER)
            elif self.__token.getTag() == Tag.TRUE:
                self.__check(Tag.TRUE)
            elif self.__token.getTag() == Tag.FALSE:
                self.__check(Tag.FALSE)
            elif self.__token.getTag() == ord('('):
                self.__check(ord('('))
                self.__expression()
                self.__check(ord(')'))
        else:
            raise Exception('Syntax Error')
        
    def __unaryExpression(self):
        if self.__token.getTag() in self.__firstPrimaryExpression:
            if self.__token.getTag() == ord('-'):
                self.__check(ord('-'))
            elif self.__token.getTag() == ord('!'):
                self.__check(ord('!'))
            else:
                self.__primaryExpression()
        else: 
            raise Exception('Syntax Error')
        
    def __extendedMultiplicativeExpression(self):
        if self.__token.getTag() in self.__firstExtendedMultiplicativeExpression:
            if self.__token.getTag() == ord('*'):
                self.__check(ord('*'))
                self.__unaryExpression()
                self.__extendedMultiplicativeExpression()
            elif self.__token.getTag() == ord('/'):
                self.__check(ord('/'))
                self.__unaryExpression()
                self.__extendedMultiplicativeExpression()
            elif self.__token.getTag() == Tag.MOD:
                self.__check(Tag.MOD)
                self.__unaryExpression()
                self.__extendedMultiplicativeExpression()
        #else:

    def __multiplicativeExpression(self):
        if self.__token.getTag() in self.__firstMultiplicativeExpression:
            self.__unaryExpression()
            self.__extendedMultiplicativeExpression()
        else: 
            raise Exception('Syntax Error')
        
    def __extendedAdditiveExpression(self):
        if self.__token.getTag() in self.__firstExtendedAdditiveExpression:
            if self.__token.getTag() == ord('+'):
                self.__check(ord('+'))
                self.__multiplicativeExpression()
                self.__extendedAdditiveExpression()
            elif self.__token.getTag() == ord('-'):
                self.__check(ord('-'))
                self.__multiplicativeExpression()
                self.__extendedAdditiveExpression()

    def __additiveExpression(self):
        if self.__token.getTag() in self.__firstAdditiveExpression:
            self.__multiplicativeExpression()
            self.__extendedAdditiveExpression()
        else: 
            raise Exception('Syntax Error')
        
    def __extendedRelationalExpression(self):
        if self.__token.getTag() in self.__firstExtendedRelationExpresion:
            if self.__token.getTag() == ord('<'):
                self.__check(ord('<'))
                self.__additiveExpression()
                self.__extendedRelationalExpression()
            elif self.__token.getTag() == Tag.LEQ:
                self.__check(Tag.LEQ)
                self.__additiveExpression()
                self.__extendedRelationalExpression()
            elif self.__token.getTag() == ord('>'):
                self.__check(ord('>'))
                self.__additiveExpression()
                self.__extendedRelationalExpression()
            elif self.__token.getTag() == Tag.GEQ:
                self.__check(Tag.GEQ)
                self.__additiveExpression()
                self.__extendedRelationalExpression()

    def __relationalExpression(self):
        if self.__token.getTag() in self.__firstRelationalExpression:
            self.__additiveExpression()
            self.__extendedRelationalExpression()
        else: 
            raise Exception('Syntax Error')
        
    def __extendedEqualityExpression(self):
        if self.__token.getTag() in self.__firstExtendedEqualityExpression:
            if self.__token.getTag() == ord('='):
                self.__check(ord('='))
                self.__relationalExpression()
                self.__extendedEqualityExpression()
            elif self.__token.getTag() == Tag.NEQ:
                self.__check(Tag.NEQ)
                self.__relationalExpression()
                self.__extendedEqualityExpression()

    def __equalityExpression(self):
        if self.__token.getTag() in self.__firstEqualityExpression:
            self.__relationalExpression()
            self.__extendedEqualityExpression()
        else:
            raise Exception('Syntax Error')
        
    def __extendedConditionalTerm(self):
        if self.__token.getTag() == Tag.AND:
            self.__check(Tag.AND)
            self.__equalityExpression()
            self.__extendedConditionalTerm()

    def __conditionalTerm(self):
        if self.__token.getTag() in self.__firstEqualityExpression:
            self.__equalityExpression()
            self.__extendedConditionalTerm()
        else:
            raise Exception('Syntax Error')
        
    def __extendedConditionalExpression(self):
        if self.__token.getTag() == Tag.OR:
            self.__check(Tag.OR)
            self.__conditionalTerm()
            self.__extendedConditionalExpression()

    def __conditionalExpression(self):
        if self.__token.getTag() in self.__firstConditionalExpression:
            self.__conditionalTerm()
            self.__extendedConditionalExpression()
        else:
            raise Exception('Syntax Error')
        
    def __expression(self):
        if self.__token.getTag() in self.__firstExpression:
            self.__conditionalExpression()
        else:
            raise Exception('Syntax Error')
        
    def __ifElseStatement(self):
        if self.__token.getTag() == Tag.IFELSE:
            self.__check(Tag.IFELSE)
            self.__expression()
            self.__check(ord('['))
            self.__statementSequence()
            self.__check(ord(']'))
            self.__check(ord('['))
            self.__statementSequence()
            self.__check(ord(']'))
        else:
            raise Exception('Syntax Error')
        
    def __ifStatement(self):
        if self.__token.getTag() == Tag.IF:
            self.__check(Tag.IF)
            self.__expression()
            self.__check(ord('['))
            self.__statementSequence()
            self.__check(ord(']'))
        else:
            raise Exception('Syntax Error')

    def __conditionalStatement(self):
        if self.__token.getTag() in self.__firstConditionalStatement:
            if self.__token.getTag() == Tag.IF:
                self.__ifStatement()
            elif self.__token.getTag() == Tag.IFELSE:
                self.__ifElseStatement()
        else:
            raise Exception('Syntax Error')

    def __repetitiveStatement(self):
        if self.__token.getTag() == Tag.REPEAT:
            self.__check(Tag.REPEAT)
            self.__expression()
            self.__check(ord('['))
            self.__statementSequence()
            self.__check(ord(']'))
        else:
            raise Exception('Syntax Error')
        
    def __structuredStatement(self):
        if self.__token.getTag() in self.__firstStructuredStatement:
            if self.__token.getTag() in self.__firstConditionalStatement:
                self.__conditionalStatement()
            elif self.__token.getTag() == Tag.REPEAT:
                self.__repetitiveStatement()
        else:
            raise Exception('Syntax Error')
        
    def __element(self):
        if self.__token.getTag() in self.__firstElement:
            if self.__token.getTag() == Tag.STRING:
                self.__check(Tag.STRING)
            elif self.__token.getTag() in self.__firstExpression:
                self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __elementList(self):
        if self.__token.getTag() == ord(','):
            self.__check(ord(','))
            self.__element()
            self.__elementList()
        
    def __textStatement(self):
        if self.__token.getTag() == Tag.PRINT:
            self.__check(Tag.PRINT)
            self.__check(ord('['))
            self.__element()
            self.__elementList()
            self.__check(ord(']'))

    def __penWidthStatement(self):
        if self.__token.getTag() == Tag.PENWIDTH:
            self.__check(Tag.PENWIDTH)
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __colorStatement(self):
        if self.__token.getTag() == Tag.COLOR:
            self.__check(Tag.COLOR)
            self.__expression()
            self.__check(ord(','))
            self.__expression()
            self.__check(ord(','))
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __penDownStatement(self):
        if self.__token.getTag() == Tag.PENDOWN:
            self.__check(Tag.PENDOWN)
        else:
            raise Exception('Syntax Error')
        
    def __penUpStatement(self):
        if self.__token.getTag() == Tag.PENUP:
            self.__check(Tag.PENUP)
        else:
            raise Exception('Syntax Error')
        
    def __arcStatement(self):
        if self.__token.getTag() == Tag.ARC:
            self.__check(Tag.ARC)
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __circleStatement(self):
        if self.__token.getTag() == Tag.CIRCLE:
            self.__check(Tag.CIRCLE)
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __clearStatement(self):
        if self.__token.getTag() == Tag.CLEAR:
            self.__check(Tag.CLEAR)
        else:
            raise Exception('Syntax Error')

    def __drawingStatement(self):
        if self.__token.getTag() in self.__firstDrawingStatement:
            if self.__token.getTag() == Tag.CLEAR:
                self.__clearStatement()
            elif self.__token.getTag() == Tag.CIRCLE:
                self.__circleStatement()
            elif self.__token.getTag() == Tag.ARC:
                self.__arcStatement()
            elif self.__token.getTag() == Tag.PENUP:
                self.__penUpStatement()
            elif self.__token.getTag() == Tag.PENDOWN:
                self.__penDownStatement()
            elif self.__token.getTag() == Tag.COLOR:
                self.__colorStatement()
            elif self.__token.getTag() == Tag.PENWIDTH:
                self.__penWidthStatement()

    def __setXYStatement(self):
        if self.__token.getTag() == Tag.SETXY:
            self.__check(Tag.SETXY)
            self.__expression()
            self.__check(ord(','))
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __setXStatement(self):
        if self.__token.getTag() == Tag.SETX:
            self.__check(Tag.SETX)
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __setYStatement(self):
        if self.__token.getTag() == Tag.SETY:
            self.__check(Tag.SETY)
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __leftStatement(self):
        if self.__token.getTag() == Tag.LEFT:
            self.__check(Tag.LEFT)
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __rightStatement(self):
        if self.__token.getTag() == Tag.RIGHT:
            self.__check(Tag.RIGHT)
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __backwardStatement(self):
        if self.__token.getTag() == Tag.BACKWARD:
            self.__check(Tag.BACKWARD)
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __forwardStatement(self):
        if self.__token.getTag() == Tag.FORWARD:
            self.__check(Tag.FORWARD)
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __movementStatement(self):
        if self.__token.getTag() in self.__firstMovementStatement:
            if self.__token.getTag() == Tag.FORWARD:
                self.__forwardStatement()
            elif self.__token.getTag() == Tag.BACKWARD:
                self.__backwardStatement()
            elif self.__token.getTag() == Tag.RIGHT:
                self.__rightStatement()
            elif self.__token.getTag() == Tag.LEFT:
                self.__leftStatement()
            elif self.__token.getTag() == Tag.SETX:
                self.__setXStatement()
            elif self.__token.getTag() == Tag.SETY:
                self.__setYStatement()
            elif self.__token.getTag() == Tag.SETXY:
                self.__setXYStatement()
            elif self.__token.getTag() == Tag.HOME:
                self.__check(Tag.HOME)
        else:
            raise Exception('Syntax Error')
        
    def __assigmentStatement(self):
        if self.__token.getTag() == Tag.ID:
            self.__check(Tag.ID)
            self.__check(Tag.ASSIGN)
            self.__expression()
        else:
            raise Exception('Syntax Error')
        
    def __identifierList(self):
        if self.__token.getTag() == ord(','):
            self.__check(ord(','))
            self.__check(Tag.ID)
            self.__identifierList()

    def __declarationStatement(self):
        if self.__token.getTag() == Tag.VAR:
            self.__check(Tag.VAR)
            self.__check(Tag.ID)
            self.__identifierList()
        else:
            raise Exception('Syntax error in declaration statement')
        
    def __simpleStatement(self):
        if self.__token.getTag() in self.__firstSimpleStatement:
            if self.__token.getTag() == Tag.VAR:
                self.__declarationStatement()
            elif self.__token.getTag() == Tag.ID:
                self.__assigmentStatement()
            elif self.__token.getTag() in self.__firstMovementStatement:
                self.__movementStatement()
            elif self.__token.getTag() in self.__firstDrawingStatement:
                self.__drawingStatement()
            elif self.__token.getTag() == Tag.PRINT:
                self.__textStatement()
        else:
            raise Exception('Syntax Error')
        
    def __statement(self):
        if self.__token.getTag() in self.__firstStatement:
            if self.__token.getTag() in self.__firstSimpleStatement:
                self.__simpleStatement()
            elif self.__token.getTag() in self.__firstStructuredStatement:
                self.__structuredStatement()
        else:
            raise Exception('Syntax Error')
        
    def __statementSequence(self):
        if self.__token.getTag() in self.__firstStatementSequence:
            self.__statement()
            self.__statementSequence()

    def __program(self):
        if self.__token.getTag() in self.__firstProgram:
            self.__statementSequence()
